### DISCUSSION TITLE
Can Windows/Non-Core Powershell 5 bugs be reported here?

### DISCUSSION THREAD

**Discussion by roke-julian-lockhart** [PowerShell/PowerShell#24299] [Q&A] [answered]
If not, I imagine that Feedback Hub is correct?

**Comment by ThomasNieto** [comment]
You are correct use feedback hub to report Windows PowerShell issues. This repository is only for cross platform PowerShell.

**Reply by roke-julian-lockhart** [reply]
> You are correct use feedback hub to report Windows PowerShell issues.

@mklement0, @ThomasNieto and @kilasuit, in that case, which forum is used? I ask because https://feedbackportal.microsoft.com/feedback/post/ doesn't include an entry for PowerShell.

**Reply by mklement0** [reply]
With said link it's also unclear to me how to target PowerShell.

However, If you use the Feedback Hub _app_ on Windows - which you can start with `Start-Process feedback-hub:` - there _is_ a way: category `Apps`, subcategory `PowerShell 7-x64`:

![image](https://github.com/user-attachments/assets/a00db7d3-3421-44af-826e-378a042460e8)

Note:

* Seemingly, based on what is installed _on my machine_, a `powershell 7` subcategory entry shows up, too  - which arguably makes more  sense than the more narrow `PowerShell 7-x64`

* However, searching among the existing issues it seems that only the `PowerShell 7-x64` subcategory has entries.

**Comment by kilasuit** [comment]
Adding to the above, please report them to feedback hub, however I'd suggest also sharing the feedback here (particularly the shareable feedback link) for others to be able to go and upvote them in the feedback hub too.

However in the [PowerShell/PowerShell repo](https://github.com/PowerShell/PowerShell) any issues raised for anything below the supported versions of PowerShell (not 5.1 as bundled in Windows) will be closed likely with `Resolution-External` tag

**Comment by mklement0** [comment, marked as answer]
#23981 is a suggestion to provide better up-front guidance about with respect to whether an issue should be reported here or not, but it is a good point that, for issues that belong in the [Windows Feedback Hub](https://support.microsoft.com/windows/send-feedback-to-microsoft-with-the-feedback-hub-app-f59187f8-8739-22d6-ba93-f66612949332) rather than in this repo, it is helpful to also have a _signpost_ issue in this repo with a shareable link to the former.

@rhubarb-geek-nz came up with a simple (but not comprehensive) test for whether a given command is being maintained in this repo (assuming you're running _PowerShell 7_ - all problems specific to _Windows PowerShell_ (the legacy edition whose latest and _last_ version is 5.1) categorically belong in the Feedback Hub):

If the following command template reports a path that **starts with `C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules`** for a given command, the module _ships with Windows_ rather than as part of PowerShell 7, so the **Feedback Hub is the right place** to report it.

```powershell
# If this reports a path that starts with "C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules", use the Feedback Hub.
(Get-Command <command-in-question>).Module.ModuleBase
```

If the **path starts with `"$PSHOME\Modules"`**, the command **_may_ be part of this repo**, but there are a few exceptions: there are **_some_ modules that are _bundled_ with PowerShell 7**, but are **maintained in _separate GitHub repos_,** and issues should be posted _there_.

As of PowerShell 7.4.x, these are:

* [`PSReadLine`](https://github.com/PowerShell/PSReadLine)
* [`ThreadJob`](https://github.com/Powershell/ThreadJob), though the plan is to replace it with [`Microsoft.PowerShell.ThreadJob`](https://github.com/PowerShell/ThreadJob) - see https://github.com/PowerShell/ThreadJob/issues/30
* [`Microsoft.PowerShell.PSResourceGet`](https://github.com/PowerShell/PSResourceGet)
* Modules that are still bundled, but no longer in active development:
  * [`Microsoft.PowerShell.Archive`](https://github.com/PowerShell/Microsoft.PowerShell.Archive)
  * Superseded by `Microsoft.PowerShell.PSResourceGet`:
    * [`PackageManagement` (aka OneGet)](https://github.com/oneget/oneget)
    * [`PowerShellGet`](https://github.com/PowerShell/PowerShellGet) (the bundled version is v2.x - see https://learn.microsoft.com/en-us/powershell/module/powershellget for background info)

If the path starts with **neither `C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules` nor `$PSHOME\Modules`**, the command is **_not_ maintained in this repo**, with the exception of a few built-in _functions_, such as `help` and `oss` (which aren't inside modules).

---

Below is the source code of **_convenience function_ `Get-CommandOrigin`**, which **automates the discovery process** described above:

Once defined, run the following, for instance (on Windows):

```powershell
PS> Get-CommandOrigin New-ScheduledTaskSettingsSet

Command         : New-ScheduledTaskSettingsSet
Type            : Function
Module          : ScheduledTasks
ModuleLocation  : C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules\ScheduledTasks
RepositoryUri   : 
Origin          : ShipsWithWindows
UsesWinPsCompat : False
```

The output can guide users as follows:

* If the `Origin` property value is `BuiltIn`, this repository _is_ the right place to file an issue.
* Otherwise:
  * If it is `Bundled`, it is a module that is bundled with PowerShell, but maintained in a separate repository, whose URL is reported in the `RepositoryUri` property; report the issue there.
  * If it is `ShipsWithWindows`, use the Feedback Hub to report the issue.
  * Otherwise (`External`), report the issue to the module's authors, using the (hopefully filled-in) `RepositoryUri` value.

`Get-CommandOrigin` source code (can be pasted at the prompt):

```powershell
function Get-CommandOrigin {
  param([Parameter(Mandatory)] $Command)
  Set-StrictMode -Version 1
  enum CommandOrigin { BuiltIn = 0; Bundled; ShipsWithWindows; External }
  $isWinPs = -not (Test-Path Variable:IsCoreCLR)
  $cmd = Get-Command $Command -ErrorAction Stop
  if ($cmd.ResolvedCommand) { $cmd = $cmd.ResolvedCommand }
  # !! This shouldn't be necessary, but it seems that for some aliases .ResolvedCommand
  # !! is unexpectedly $null even though the target command exists.
  while ($cmd -is [System.Management.Automation.AliasInfo]) {
    $cmd = Get-Command $cmd.Definition -ErrorAction Stop
  }
  # See if the module is located inside the $PSHOME directory, which implies either a built-in or a bundled module.
  $builtInOrBundled = $cmd.Module -and $cmd.Module.Path.StartsWith($PSHOME + [IO.Path]::DirectorySeparatorChar, 'InvariantCultureIgnoreCase')
  # Built-in modules only have *.psd1 files and possibly formatting / type-definition files;
  # Conversely, the presence a *.dll or *.psm1 file implies a *bundled* module, such as PSReadLine.
  # Exception: One Microsoft.PowerShell.* module, Microsoft.PowerShell.Core, is built into the binary
  #            and has no external representation in the file-system. It can be identified by having a
  #            having a $null-valued .Module property while having a non-empty .ModuleName property.
  $builtIn = (-not $cmd.Module -and $cmd.ModuleName) -or 
             ($builtInOrBundled -and -not (Get-ChildItem -Path (Join-Path (Split-Path $cmd.Module.Path) *) -Include *.dll, *.psm1))
             # !! To work around a bug in WinPS where using -LiteralPath causes -Include to be ignored, we use -Path + \* instead.
             # !! For built-in modules, this should be safe.
  $bundled = $builtInOrBundled -and -not $builtIn
  $usesWinPsCompat = $env:OS -eq 'Windows_NT' -and $cmd.Module.Path -match '\\remoteIpMoProxy_'
  # Note: !! .ModuleBase, which is the manifest file's directory path only, incorrectly reports
  #       !! $PSHOME for modules loaded via the Windows PowerShell compatibility feature.
  #       !! Use of .Path avoids this problem.
  #       !! In the past, we also tried to use .ModuleBase to avoid a problem with .Path reporting a path
  #       !! to a DLL in the GAC rather than a module directory among the WinPS system modules.
  #       !! It's unclear what modules may have been affected, but to be safe we test for just $env:SystemRoot at the start.
  $shipsWithWindows = $usesWinPsCompat -or
    (-not $builtIn -and $env:OS -eq 'Windows_NT' -and $cmd.Module -and
     $cmd.Module.Path.StartsWith("$env:SystemRoot\", 'InvariantCultureIgnoreCase'))
  # Also check for built-in *functions* defined directly in the global scope domain, such as `oss` and `help`
  # Note: The list of built-in function is *hard-coded*, and current as of PowerShell 7.4.x
  if (-not $builtIn -and $cmd.CommandType -eq 'Function' -and -not $cmd.Module) {
    [bool] $isCore = Get-Variable -ValueOnly IsCoreCLR -ErrorAction Ignore
    $builtInFuncs = 'cd..', 'cd\', 'Clear-Host', 'help', 'oss', 'Pause', 'prompt', 'TabExpansion2'
    if ($isCore) { $builtInFuncs += 'cd~' }
    if ($env:OS -eq 'Windows_NT') { $builtInFuncs += @('mkdir') + ([char[]] (65..90) -replace '$', ':') }
    $builtIn = $cmd.Name -in $builtInFuncs
  }
  $repoUri = if ($builtIn) { 'https://github.com/PowerShell/PowerShell' } else { [string] $cmd.Module.ProjectUri }
  # The bundled ThreadJob and Microsoft.PowerShell.Archive modules don't define a project URI; we supply it manually.
  # For ThreadJob, the *new* repo for the Microsoft.PowerShell.ThreadJob module is reported, which will be used going forward.
  # Note that Microsoft.PowerShell.Archive is no longer actively deloped.
  if (-not $repoUri -and $bundled) {
    switch ($cmd.ModuleName) {
     'ThreadJob' { $repoUri = 'https://github.com/Powershell/ThreadJob' }
     'Microsoft.PowerShell.Archive' { $repoUri = 'https://github.com/PowerShell/Microsoft.PowerShell.Archive' }
    }
  }
  # Construct and output an object with the results.
  [pscustomobject] @{
    Command = $cmd
    Type = $cmd.CommandType
    Module = $cmd.Module
    ModuleLocation = if ($usesWinPsCompat) { Split-Path $cmd.Module.Path } else { $cmd.Module.ModuleBase }
    RepositoryUri = $repoUri
    Origin = $(
      if ($builtIn) { [CommandOrigin]::BuiltIn }
      elseif ($bundled -and -not $isWinPs) { [CommandOrigin]::Bundled }
      elseif ($shipsWithWindows) { [CommandOrigin]::ShipsWithWindows }
      else { [CommandOrigin]::External }
    )
    UsesWinPsCompat = if ($shipsWithWindows -and -not $isWinPs) { $usesWinPsCompat }
  }
}
```

**Reply by roke-julian-lockhart** [reply]
Thank you for this, @mklement0. That's a brilliant toolset. I suggest that you package this as a module, if that would be feasible - I imagine many would find such functionality useful.
